feat(ci): Only run relevant jobs when there were changes to that file

klochowicz opened the pull request:
Apart from the main advertised thing (skipping CI jobs when nothing changed), it also removes the maintenance burden of manually updating various files when adding new crate.

includes new meta CI job that tells bors when everything passed.

klochowicz committed:
feat(ci): Only run relevant jobs when there were changes to that file

klochowicz committed:
chore: Use a dummy CI job to signal overall CI status to bors

It is extremely fragile to need to specify by hand the expanded matrix jobs from
GitHub CI. Furthermore, in case if I job gets skipped, it is not expanded,
making the bors timeout.

Follow suggestion found in bors documentation and set a job that's sole purpose
is to tell bors whether we passed the CI.

See: https://bors.tech/documentation/getting-started/

Note: As a side effect, we won't need to manually update bors.toml whenever we
add a new crate to the workspace.

luckysori reviewed (approved):
Still not thrilled about e53334878e506c17bc58aa31acf8ad94bfb268e1, but I like the other changes 😃

bonomat commented on .github/workflows/ci.yml:
```diff
@@ -18,6 +18,27 @@ jobs:
+              - 'Cargo.toml'
+              - 'Cargo.lock'
+            frontend:
+              - 'taker-frontend/**'
```
you are missing `taker-electron`

klochowicz replied:
w8 a second, `frontend` job does not build or check electron stuff, so the check should not really trigger it. 

I propose to address it by adding a separate check and job for electron CI stuff (as it was not covered here anyway, so it's not a regression)

edit: I think that you meant that electron should be rebuilt when taker-frontend changes, because we embed it.

klochowicz replied:
I've opened #3050

bonomat replied:
Oh, we actually have 0 checks in taker-electron. So there is nothing to do here :D

bonomat reviewed (approved):
I guess I'll survive the project restructuring. If not, I know where to complain 😈

holzeis commented on .github/workflows/ci.yml:
```diff
@@ -18,6 +18,27 @@ jobs:
+    runs-on: ubuntu-latest
+    # Set job outputs to values from filter step
+    outputs:
+      backend: ${{ steps.filter.outputs.backend }}
```
🔧 Why only separate between frontend and backend.. we could even be more specific.

e.g. why build the maker-frontend if only the taker-frontend has changed? etc.

I would create an change output for every sub project we have. It might be cumbersome because of dependencies, but I think that's a complexity worth maintaining.

klochowicz replied:
Happy to accept further patches! 😁 it will be covered by #3050 (that will cover electron changes, and as a prerequisite for that will have to split maker and taker frontends.)

holzeis reviewed (commented):
LGTM 👍 but I think it doesn't go far enough! This change will still build project unnecessarily.

klochowicz committed:
chore: Reorganise crates in the workspace into crates directory

This makes it simpler to add/remove crate from the workspace (no need to
maintain a list), and also makes it much easier to grok the itchysats repo,
which has many non-crate things too.

This also will allow us to easy match on changed Rust files for the CI
purposes (optimising the CI runs for when we *do not* change any Rust code).
